Create classes for Application Load Balancer

wimnat opened the pull request:
##### SUMMARY
Create a super class for ELBv2 and then a class for ALB.

This is so that the NLB module can be created and re-use a vast majority of this code.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
elb_application_lb

##### ANSIBLE VERSION
For 2.6

wimnat committed:
Create classes for Application Load Balancer

wimnat commented:
@s-hertel @willthames 

Hi guys,

A couple of questions here...

I want to add some ITs.  Do we have things like a VPC and subnets all spun up and ready to use or do we need to include those tasks along side the ALB tasks?

I've created a couple of classes here, all in one file.  They are all related but dependent so they could be in seperate files.  Is that convention or should I just carry on using one file?

If you could perform a few manual tests that'd be great.  Most of the code is the same but it's obviously quite a big refactor.  The benefit is that NLB module should be pretty straight-forward after this.

willthames commented:
Hi @wimnat, the integration tests spin up and tear down the relevant VPCs as subnets, security groups etc. are all somewhat dependent on whatever is being tested. There might be an argument to be made to provide create_vpc and destroy_vpc roles, but I'm not yet sure. 

I'd say the recently merged ec2_vpc_route_table test suite is a reasonable starting point these days of how we do set up and tear down while DRYing AWS parameters.

wimnat commented:
@s-hertel @willthames any support on the above build error?  Not sure why it's reporting it can't find boto3 module :/

s-hertel commented:
@wimnat the module_utils files need exception handling around the imports. You can just do something along the lines of:
```
try:
    import botocore
except ImportError:
    pass
```
And then as long as the modules in question are checking for HAS_BOTO3 that should work.

wimnat committed:
Fix up copyright and botocore import

wimnat committed:
add description in ec2_group tasks

wimnat commented:
@s-hertel @willthames just a note on the unit tests for this module - they fail now and i'm not quite sure how to fix them or if it's even possible with the restructure.  Would you like to remove them to get the build passing or is there someone who can help to get them fixed up?

s-hertel commented:
@wimnat Just wanted to leave an update here that I'm working on fixing up the tests and will have a commit in a day or two.

s-hertel commented:
These unit tests are a pain. I was going to write new ones using the placebo fixtures but importing the fixtures doesn't work because test/ isn't a package (the fixtures live next to the module tests rather than the module_utils ones). I am in favor of removing the now-useless unit tests since this pull request added integration tests.

@ryansb What do you think?

s-hertel commented:
cc/ @willthames too ^

wimnat commented:
Regarding integration tests - i will extend these either here or in a separate PR once the build is passing successfully.

s-hertel commented on lib/ansible/module_utils/aws/elb_utils.py:
```diff
@@ -0,0 +1,96 @@
+    try:
+        load_balancer_paginator = connection.get_paginator('describe_load_balancers')
+        return (load_balancer_paginator.paginate(Names=[elb_name]).build_full_result())['LoadBalancers'][0]
+    except ClientError as e:
```
Exception guidelines have been updated to also catch BotoCoreError: https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/GUIDELINES.md#boto3-1. BotoCoreError exceptions don't have a .response, so it will need separate handling from ClientError. Something like this:
```
except BotoCoreError as e:
    module.fail_json(msg="Unable to describe load balancers: {0}".format(to_native(e)), exception=traceback.format_exc())
```

Need to add BotoCoreError handling in other places too.

s-hertel commented on lib/ansible/module_utils/aws/elb_utils.py:
```diff
@@ -0,0 +1,96 @@
+        if e.response['Error']['Code'] == 'LoadBalancerNotFound':
+            return None
+        else:
+            module.fail_json(msg=e.message, exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
```
Python3 does not have exception.message. Instead, use a custom message to describe the failed operation and add the exception to the end, like
`module.fail_json(msg='Unable to describe load balancers: {0}'.format(to_native(e)), ...)`

Need to remove e.message in other places too.

s-hertel commented on lib/ansible/module_utils/aws/elbv2.py:
```diff
@@ -0,0 +1,737 @@
+                self.module.fail_json(msg=e.message, exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
+
+
+class ELBListeners:
```
This should be a new style class - `class ELBListeners(object):`

s-hertel commented on lib/ansible/module_utils/aws/elbv2.py:
```diff
@@ -0,0 +1,737 @@
+            return None
+
+
+class ELBListener:
```
This should also be a new style class.

s-hertel commented on lib/ansible/module_utils/aws/elbv2.py:
```diff
@@ -0,0 +1,737 @@
+            self.module.fail_json(msg=e.message, exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
+
+
+class ELBListenerRules:
```
This should also be a new style class

s-hertel commented on lib/ansible/module_utils/aws/elbv2.py:
```diff
@@ -0,0 +1,737 @@
+        return rules_to_add, rules_to_modify, rules_to_delete
+
+
+class ELBListenerRule:
```
This should also be a new style class

s-hertel commented on lib/ansible/modules/cloud/amazon/elb_application_lb.py:
```diff
@@ -969,7 +489,7 @@ def main():
             access_logs_s3_bucket=dict(type='str'),
             access_logs_s3_prefix=dict(type='str'),
-            deletion_protection=dict(default=False, type='bool'),
+            deletion_protection=dict(type='bool'),
```
This is documented to default to False. Is that a bug? Should update the docs too.

wimnat replied:
Yep this is an existing bug.  Any default of this type means that if you leave off the parameter in the module then it could end up being changed from what is configured.  E.g. if a user manually turned on deletion protection and then reran the module without the parameter specified, Ansible will turn it off.

I'll update the doc.

wimnat committed:
Ensure all classes are new type. Update error handling to use fail_json_aws

s-hertel commented:
@wimnat I talked to @ryansb about the unit tests. He's +1 to remove in favor of replacing with better integration tests. This is looking good to me.

wimnat committed:
Remove ALB unit tests

wimnat committed:
fix alb name for ITs

wimnat commented:
@s-hertel thank you! Progress :)

I have removed the unit tests in this branch but now I'm faced with a permission error for my ITs - no permission to create a Load Balancer.  Is this something you can fix in the backend?  I had a quick search through the ansible repo for an IAM policy assumed by shippable but couldn't see anything.

linuxdynasty commented:
@wimnat why do you use the AWSRetry decorator around the methods in this library. Currently I use Ansible for all of the AWS provisioning and I tend to have to have all of the AWS modules I use locally and modify them all to use the AWSRetry decorator to get around the AWS Throttling error.

Example below

```
@ec2.AWSRetry.backoff()                                                                                                                                                                                                                                                [51/8280]
def cluster_status(client, cluster_name):
    success = True
    err_msg = ''
    status = ''
    try:
        status = (
            client.describe_clusters(
                ClusterIdentifier=cluster_name,
            )['Clusters'][0]['ClusterStatus']
        )

    except Exception as e:
        success = False
        err_msg = e.response['message']

    return (success, err_msg, status)
```

That being said if you choose not to use it, I will still use it locally, but it would be more convenient for users who run into the throttling errors that I use to run into before implementing that module.

wimnat commented:
@linuxdynasty would you mind creating a seperate issue with this and tagging me?  There are a couple of outstanding issues with ALB module that i would like to address but not as part of this PR (as it's already so big).  A seperate issue will help me remember :)

wimnat commented:
@s-hertel bump!

wimnat commented:
@s-hertel @willthames i've removed the WIP from this.  I just need to get the ITs passing (need IAM permission) and then i think it's good for merge.  There are some other issues out on this module that i'd like to address once this is in.

wimnat commented:
@willthames @ryansb is the a particular reason this and subsequently the NLB module didn't make the cut for 2.5?

s-hertel commented:
Lack of time, not for lack of wanting it. NLB was on the roadmap for 2.5 and it slipped. This is such a big change, it's better that it gets a proper review.  I'm going to push some modifications to the tests so they can run successfully in CI (hardcoded names will cause problems when running in parallel) and an IGW isn't being created. I have opened a pull request to add CI permissions.

s-hertel committed:
Fix running tests in parallel by using resource_prefix in resource names

Create an internet gateway

Fix always section so it runs

s-hertel commented:
These tests will need to be run manually. Adding a commit to remove the CI alias. My CI PR was declined because there is no way to restrict resource-level permissions for these tests.
From this page https://docs.aws.amazon.com/elasticloadbalancing/latest/userguide/load-balancer-authentication-access-control.html:
```
There are two API sets for Elastic Load Balancing: one for Application Load Balancers and Network Load Balancers (API version 2015-12-01), and one for Classic Load Balancers (API version 2012-06-01). Currently, the 2015-12-01 API does not support resource-level permissions. The version 2012-06-01 API supports resource-level permissions; however, not every API action supports resource-level permissions.
```

s-hertel committed:
Remove CI alias since these cannot be safely run in CI

s-hertel commented on lib/ansible/module_utils/aws/elbv2.py:
```diff
@@ -0,0 +1,737 @@
+        self.deletion_protection = module.params.get("deletion_protection")
+        self.wait = module.params.get("wait")
+
+        if module.params.get("tags"):
```
Should this be `if module.params.get("tags") is not None:`? That way, users can remove their tags by giving an empty dict and not have self.tags set to None.

wimnat replied:
Yep 👍

s-hertel commented on lib/ansible/module_utils/aws/elbv2.py:
```diff
@@ -0,0 +1,737 @@
+
+        self.connection_ec2 = connection_ec2
+
+        # Ansible module parameters specific to ALBs
```
*specific to NLBs

wimnat replied:
Done

s-hertel commented on lib/ansible/module_utils/aws/elbv2.py:
```diff
@@ -0,0 +1,737 @@
+        else:
+            self.elb_attributes = None
+
+    def wait_for_status(self, elb_name, status):
```
There are some waiters we could probably leverage instead: http://boto3.readthedocs.io/en/latest/reference/services/elbv2.html#waiters

s-hertel commented on lib/ansible/module_utils/aws/elbv2.py:
```diff
@@ -0,0 +1,737 @@
+                else:
+                    time.sleep(polling_increment_secs)
+            except (BotoCoreError, ClientError) as e:
+                self.module.fail_json_aws(e, msg=str(e))
```
Note for this and subsequent uses of module.fail_json_aws - it uses e in the message output already so if msg is just e it is redundant. Better have the message output something helpful to the user about where/what failed/what parameters were used or not have a msg at all.

wimnat replied:
I've noticed that in a few modules already.  I'm not a big fan of the custom message because ClientError could indicate so many things.  Sure, most of the time the error thrown will match up with the custom message but i'm sure there could be cases where the custom message does not really relate to the Exception unless you're very careful.  I can see it causing much confusion.

Anyway, point noted about not requiring if i have no custom message so i will remove

s-hertel commented on lib/ansible/module_utils/aws/elbv2.py:
```diff
@@ -0,0 +1,737 @@
+        if self.security_groups is not None:
+            params['SecurityGroups'] = self.security_groups
+        params['Scheme'] = self.scheme
+        if self.tags is not None:
```
Right now this works but if my suggestion is taken above, this should just be `if self.tags` since if tags is an empty list it shouldn't be used in the create call, only when removing rules.

s-hertel commented on lib/ansible/module_utils/aws/elbv2.py:
```diff
@@ -0,0 +1,737 @@
+                self.connection.modify_load_balancer_attributes(LoadBalancerArn=self.elb['LoadBalancerArn'], Attributes=update_attributes)
+                self.changed = True
+            except (BotoCoreError, ClientError) as e:
+                # Something went wrong setting attributes. If this ELB was created during this task, delete it to leave a consistent state
```
Not sure about deleting the ELB here. If the delete_load_balancer call fails the user will never see this failure. Since the module is idempotent the user should be able to rerun the playbook after fixing whatever was wrong (if it was a param error, otherwise just rerunning the playbook). But if you think this is right I think the failure message should definitely indicate what happened.

wimnat replied:
I chose this behaviour because if the ELB is left there then the user will start paying for it and it may not be expected considering they saw the task fail.

When you say you think the failure message should indicate what happened do you mean wrap the delete_load_balancer call in another `try`?  Or just a generic message saying 'We created the load balancer but failed to set attributes so we removed it'?

s-hertel commented on lib/ansible/module_utils/aws/elbv2.py:
```diff
@@ -0,0 +1,737 @@
+                # Something went wrong setting attributes. If this ELB was created during this task, delete it to leave a consistent state
+                if self.new_load_balancer:
+                    self.connection.delete_load_balancer(LoadBalancerArn=self.elb['LoadBalancerArn'])
+                self.module.fail_json_aws(e, msg=str(e))
```
Same qualm as above. This behavior should be noted in the failure message.

s-hertel commented on lib/ansible/module_utils/aws/elbv2.py:
```diff
@@ -0,0 +1,737 @@
+
+        return listeners_to_add, listeners_to_modify, listeners_to_delete
+
+    def _compare_listener(self, current_listener, new_listener):
```
Since current_listener and new_listener are dictionaries, you could make them hashable for easy comparison. The compare_policies function in module_utils.ec2 is being used for that in the aws_waf_condition module.
```
from ansible.module_utils.ec2 import compare_policies as compare_dicts
...
def _compare_listener(self, current_listener, new_listener):
    if compare_dicts(current_listener, new_listener):
        # the dictionaries are different
        return new_listener
    else:
        return None
```

s-hertel commented on lib/ansible/module_utils/aws/elbv2.py:
```diff
@@ -0,0 +1,737 @@
+        modified_rule = {}
+
+        # Priority
+        if current_rule['Priority'] != new_rule['Priority']:
```
This function may also be able to use compare_policies to cut down on the logic.

s-hertel commented on lib/ansible/modules/cloud/amazon/elb_application_lb.py:
```diff
@@ -342,623 +341,137 @@
+    for listener_to_delete in listeners_to_delete:
+        listener_obj = ELBListener(elb_obj.connection, elb_obj.module, listener_to_delete, elb_obj.elb['LoadBalancerArn'])
+        listener_obj.delete()
+        listeners_obj.changed = True
```
Should this be set in listener_obj.delete()?

wimnat replied:
See response to similar comment below

s-hertel commented on lib/ansible/modules/cloud/amazon/elb_application_lb.py:
```diff
@@ -342,623 +341,137 @@
-            module.fail_json(msg=e.message, exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
+        listener_obj = ELBListener(elb_obj.connection, elb_obj.module, listener_to_add, elb_obj.elb['LoadBalancerArn'])
+        listener_obj.add()
+        listeners_obj.changed = True
```
Is there a reason not to set this in listener_obj.add()?

wimnat replied:
`listener_obj` is a seperate class from `listeners_obj` and it does not have a `changed` parameter.  We don't really care if a listener has changed, only if the listeners as a whole have changed hence why it's setting on the listeners_obj object.

s-hertel commented on lib/ansible/modules/cloud/amazon/elb_application_lb.py:
```diff
@@ -342,623 +341,137 @@
-            module.fail_json(msg=e.message, exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
+        listener_obj = ELBListener(elb_obj.connection, elb_obj.module, listener_to_modify, elb_obj.elb['LoadBalancerArn'])
+        listener_obj.modify()
+        listeners_obj.changed = True
```
Here too?

wimnat replied:
As above

s-hertel commented on lib/ansible/modules/cloud/amazon/elb_application_lb.py:
```diff
@@ -1006,6 +519,22 @@ def main():
                 elif key == 'Port':
                     listener[key] = int(listener[key])
 
+                if key == 'Protocol' and listener[key] == 'HTTPS':
```
These validations can use suboptions in the argument_spec instead. Here's an example of a module using options: https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/azure/azure_rm_keyvault.py#L227-L238

wimnat replied:
Done, but i still had to leave some `if` validation in because it looks like `required_if` does not support this new suboptions stuff.

I also kept the options as CamelCase to maintain backward compatibility

s-hertel commented on lib/ansible/modules/cloud/amazon/elb_application_lb.py:
```diff
@@ -1006,6 +519,22 @@ def main():
+            if 'Port' not in listener.keys():
+                module.fail_json(msg="'Port' is a required listener dict key")
+
     if not HAS_BOTO3:
```
Now that the module uses AnsibleAWSModule you can remove this and the import.

wimnat replied:
Done

s-hertel committed:
Add a few more integration tests for elb_application_lb

wimnat committed:
Fixes based on PR review

wimnat commented:
@s-hertel thanks for the review.  I've made some changes based on your comments.

And thanks for your explanation regarding`2.5`... I know things slip.  Such is life.  I just wanted to make sure there wasn't a road block on my part.

wimnat committed:
validation fixes

s-hertel committed:
Remove listener options that are None since they default to that now that they use suboptions

Fix suboption types

Fix checking HTTPS protocol options are provided

Update connection creation since AnsibleAWSModule is used

Add tests for invalid listener options

Fix removing tags and uncomment tests for that

Add unsupported CI alias

Remove redundant msg=str(e) from fail_json_aws

s-hertel committed:
Document wait_timeout

Fix dict comprehension

s-hertel committed:
Add AWSRetry

wimnat commented on lib/ansible/module_utils/aws/elb_utils.py:
```diff
@@ -13,6 +15,22 @@ def get_elb(connection, module, elb_name):
+        module.fail_json_aws(e)
+
+
+@AWSRetry.jittered_backoff()
```
@s-hertel i'm not sure this works here.  If the ELB doesn't exist then the retry logic means it'll keep retrying causing a large delay even though it's quite likely we just want to move on and create the ELB.

s-hertel replied:
@wimnat I think this is right. It should return None without retrying at all if the Load Balancer doesn't exist. It will just retry for exceptions that aren't handled. I'm checking this with a concrete example now though.

Yes. I turned on the timer callback and used this playbook:
```
---
- hosts: localhost
  connection: local
  gather_facts: no
  tasks:
    - name: create a new application lb
      elb_application_lb:
        profile: shertel
        region: ca-central-1
        name: "{{ lb_name }}"
        security_groups:
          - "{{ sg_1 }}"
        subnets:
          - "{{ subnet_1 }}"
          - "{{ subnet_2 }}"
        state: present

    - name: clean up
      elb_application_lb:
        profile: shertel
        region: ca-central-1
        name: "{{ lb_name }}"
        state: absent
```
These tasks take 2 seconds for me on devel and on this branch.
`Playbook run took 0 days, 0 hours, 0 minutes, 2 seconds`

s-hertel replied:
If I remove the `return None` and raise the exception again instead it does retry (before failing since the `return None` is necessary): `Playbook run took 0 days, 0 hours, 4 minutes, 33 seconds`

wimnat replied:
Hmm I had different behaviour over the weekend. Let me retry and get back to you

wimnat replied:
Hmm I had different behaviour over the weekend. Let me retry and get back to you

wimnat replied:
Can not replicate behaviour so let's sweep this under the carpet ;)

willthames commented on test/integration/targets/elb_application_lb/tasks/main.yml:
```diff
@@ -0,0 +1,524 @@
+        az: "b"
+        public: "False"
+      - cidr: 10.228.231.128/25
+        az: "c"
```
Do we need to use this third zone (I'm trying to keep the tests region independent and I know that not all regions have more than two zones)

wimnat replied:
Removed :)

willthames commented on lib/ansible/module_utils/aws/elbv2.py:
```diff
@@ -0,0 +1,757 @@
+                self.connection.describe_tags
+            )(ResourceArns=[self.elb['LoadBalancerArn']])['TagDescriptions'][0]['Tags']
+        except (BotoCoreError, ClientError) as e:
+            self.module.fail_json_aws(e)
```
Can we have helpful messages when failing? I tend to like `module.fail_json_aws(e, "Couldn't get tags")` but perhaps I'm being fussy and perhaps it's obvious from the stacktrace.

wimnat replied:
I always find AWS messages good enough and a lot of the other modules i've written rely on that.  I haven't seen any noise in the issues to make me think that other users don't think the same.  Happy to accomodate though.

willthames commented on lib/ansible/module_utils/aws/elbv2.py:
```diff
@@ -0,0 +1,757 @@
+                    get_ec2_security_group_ids_from_names
+                )(module.params.get('security_groups'), self.connection_ec2, boto3=True)
+            except ValueError as e:
+                self.module.fail_json(msg=str(e), exception=traceback.format_exc())
```
Not sure that this couldn't just use module.fail_json_aws(e)

wimnat replied:
I'm not familiar enough with fail_json_aws but this exception wont be from AWS so i'm assuming not.  Could be wrong.

willthames commented on lib/ansible/module_utils/aws/elbv2.py:
```diff
@@ -0,0 +1,757 @@
+        self.access_logs_s3_prefix = module.params.get("access_logs_s3_prefix")
+        self.idle_timeout = module.params.get("idle_timeout")
+
+    def create_elb(self):
```
This looks fairly similar for both kinds of load balancer, other than the security_groups - could this be lifted into the parent class?

wimnat replied:
Yes, it probably could, but for future-proofing i think it's best to leave here.  Who knows how the two types may diverge in the future or even if there will be a third or fourth type that has very different options.

wimnat committed:
Removed az-c from tests as not all regions have 3 zones

wimnat committed:
Add type:bool for wait option and update ignore.txt

s-hertel committed:
add back cloud alias

s-hertel committed:
remove from validate modules ignore list

s-hertel committed:
Fix tests now that there are only two subnets for compatibility in all regions

s-hertel committed:
Add tests using the ALB

Restructure the tests

s-hertel committed:
move tests into same directory
